src/components/views/dialogs/keybackup/CreateKeyBackupDialog.js

```diff
@@ -0,0 +1,230 @@
+/*
+Copyright 2018 New Vector Ltd
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import React from 'react';
+import sdk from '../../../../index';
+import MatrixClientPeg from '../../../../MatrixClientPeg';
+import Promise from 'bluebird';
+
+import { _t, _td } from '../../../../languageHandler';
+
+const PHASE_INTRO = 0;
+const PHASE_GENERATING = 1;
+const PHASE_SHOWKEY = 2;
+const PHASE_MAKEBACKUP = 3;
+const PHASE_UPLOAD = 4;
+const PHASE_DONE = 5;
+
+// XXX: copied from ShareDialog: factor out into utils
+function selectText(target) {
+    const range = document.createRange();
+    range.selectNodeContents(target);
+
+    const selection = window.getSelection();
+    selection.removeAllRanges();
+    selection.addRange(range);
+}
+
+/**
+ * Walks the user through the process of creating an e22 key backup
+ * on the server.
+ */
+export default React.createClass({
+    getInitialState: function() {
+        return {
+            phase: PHASE_INTRO,
+        };
+    },
+
+    componentWillMount: function() {
+        this._recoveryKeyNode = null;
+        this._keyBackupInfo = null;
+    },
+
+    _collectRecoveryKeyNode: function(n) {
+        this._recoveryKeyNode = n;
+    },
+
+    _copyRecoveryKey: function() {
+        selectText(this._recoveryKeyNode);
+        const successful = document.execCommand('copy');
+        if (successful) {
+            this.setState({copied: true});
+        }
+    },
+
+    _createBackup: function() {
+        this.setState({
+            phase: PHASE_MAKEBACKUP,
+            error: null,
+        });
+        this._createBackupPromise = MatrixClientPeg.get().createKeyBackupVersion(
+            this._keyBackupInfo,
+        ).then((info) => {
+            this.setState({
+                phase: PHASE_UPLOAD,
+            });
+            return MatrixClientPeg.get().backupAllGroupSessions(info.version);
+        }).then(() => {
+            this.setState({
+                phase: PHASE_DONE,
+            });
+        }).catch(e => {
+            console.log("Error creating key backup", e);
+            this.setState({
+                error: e,
+            });
+        });
+    },
+
+    _onCancel: function() {
+        this.props.onFinished(false);
+    },
+
+    _onDone: function() {
+        this.props.onFinished(true);
+    },
+
+    _generateKey: async function() {
+        this.setState({
+            phase: PHASE_GENERATING,
+        });
+        // Look, work is being done!
+        await Promise.delay(1200);
```

Comment:
okay, i'm totally missing why we'd want a pregnant pause at this point

Reply:
This was me making it look like it's doing something to help the user understand the key is being generated rather than just shown to you, which it sort of des if it comes up instantly. Given it's all temporary UI I thought I'd experiment. :)

Reply:
http://thedailywtf.com/articles/The-Speedup-Loop